# Heather's Anomaly Detection Project Notebook
by Heather Ortega McMillan

This is the initial notebook compiled by Heather. For full project details see the Final Notebook

In [3]:
import pandas as pd
import numpy as np

from acquire import get_joined_curriculum_data

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

In [4]:
df = get_joined_curriculum_data()

1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?
    - cohort pulled from the sql table 
    - sort all the endpoints 
    
7. Which lessons are least accessed?
    - kind of related to 2

LUPE ^^^^ 

### Question I'm tackling 
2. Is there a cohort that referred to a lesson significantly more than other cohorts seemed to gloss over?
    - divided web dev and data science (4 different programs)
    - sort all the endpoints 
    - counts of cohorts accessing same endpoints
    

    

### Program ID Key 
1 - Full stack PHP

2 - Java

3 - Data Science

4 - Front End Program


In [10]:
# deal with nulls by filling with 0s
#df.cohort_id.value_counts(dropna=False)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1018810 entries, 0 to 1018809
Data columns (total 9 columns):
 #   Column      Non-Null Count    Dtype  
---  ------      --------------    -----  
 0   date        1018810 non-null  object 
 1   endpoint    1018809 non-null  object 
 2   user_id     1018810 non-null  int64  
 3   cohort_id   965313 non-null   float64
 4   source_ip   1018810 non-null  object 
 5   name        954313 non-null   object 
 6   start_date  954313 non-null   object 
 7   end_date    954313 non-null   object 
 8   program_id  954313 non-null   float64
dtypes: float64(2), int64(1), object(6)
memory usage: 77.7+ MB


In [ ]:
# Fill nulls
def fill_nulls(df):
    '''
    This function fills the nulls in the joined curriculum log data
    '''
    df = df.fillna(value = 0)
    
    return df
    
    

In [20]:
df = df.fillna(value = 0)

In [28]:
df[df.name == 0]

date                               endpoint  user_id  cohort_id  ... name start_date end_date program_id
411      2018-01-26                                      /       48        0.0  ...    0          0        0        0.0
412      2018-01-26  spring/extra-features/form-validation       48        0.0  ...    0          0        0        0.0
425      2018-01-26                                      /       48        0.0  ...    0          0        0        0.0
435      2018-01-26                                      /       48        0.0  ...    0          0        0        0.0
436      2018-01-26              mysql/relationships/joins       48        0.0  ...    0          0        0        0.0
...             ...                                    ...      ...        ...  ...  ...        ...      ...        ...
1018759  2021-07-15                  python/intro-to-numpy     1026      165.0  ...    0          0        0        0.0
1018778  2021-07-15                  python/intro-to-numpy     1051      165.0  ...    0          0        0        0.0
1018785  2021-07-15                  python/intro-to-numpy     1037      165.0  ...    0          0        0        0.0
1018786  2021-07-15                  python/intro-to-numpy     1041      165.0  ...    0          0        0        0.0
1018802  2021-07-15                  python/intro-to-numpy     1048      165.0  ...    0          0        0        0.0

[64497 rows x 9 columns]

#### Takeaways
- Newer people with no cohort id must be from Germaine and the Web Dev cohorts that just started

In [ ]:
# Is there a cohort that referred to a lesson significantly more than other cohorts seemed to 
# gloss over?


# See Lessons (endpoint) Count of each cohort access 

df['endpoint', 'cohort']



In [31]:
df[['name', 'cohort_id']].value_counts()

name        cohort_id
Staff       28.0         96954
0           0.0          53497
Ceres       33.0         40858
Jupiter     62.0         39316
Zion        29.0         38198
Fortuna     53.0         37043
Voyageurs   24.0         35707
Deimos      51.0         35381
Ganymede    57.0         34919
Apex        56.0         34172
Marco       135.0        34093
Darden      59.0         32871
Teddy       22.0         31258
Hyperion    58.0         30137
Betelgeuse  32.0         29754
Ulysses     23.0         28581
Europa      52.0         28177
Xanadu      26.0         27811
Bayes       34.0         27261
Wrangell    25.0         25680
Andromeda   31.0         25372
Kalypso     132.0        24978
Neptune     138.0        23744
Curie       55.0         22064
Florence    137.0        21435
Luna        134.0        21095
Yosemite    27.0         20864
Arches      1.0          19063
Easley      133.0        18371
Bash        61.0         18212
Oberon      139.0        17745
Lassen      14.0 

In [33]:
df[df.cohort_id == 166]

date                  endpoint  user_id  cohort_id        source_ip name start_date end_date  program_id
995542   2021-06-28                         /     1054      166.0   68.201.208.230    0          0        0         0.0
995543   2021-06-28                         /     1055      166.0  173.172.170.207    0          0        0         0.0
995544   2021-06-28                         /     1056      166.0   173.173.103.52    0          0        0         0.0
995545   2021-06-28                         /     1057      166.0    136.50.55.145    0          0        0         0.0
995546   2021-06-28                         /     1058      166.0      24.26.246.9    0          0        0         0.0
...             ...                       ...      ...        ...              ...  ...        ...      ...         ...
1018646  2021-07-15  search/search_index.json     1067      166.0    136.50.91.119    0          0        0         0.0
1018647  2021-07-15        javascript-i/loops     1061      166.0     75.51.234.24    0          0        0         0.0
1018663  2021-07-15              javascript-i     1067      166.0    136.50.91.119    0          0        0         0.0
1018664  2021-07-15        javascript-i/loops     1067      166.0    136.50.91.119    0          0        0         0.0
1018714  2021-07-15        javascript-i/loops     1055      166.0  173.172.170.207    0          0        0         0.0

[1829 rows x 9 columns]

cohort_id 165 must be Germaine program code of 3
cohort_id 166 must be new web dev cohort (Polaris) program code of 2